woocommerce/google-listings-and-ads
# Product titles for Free Listing reports

### Changes proposed in this Pull Request:

Since the Google Merchant Center API does not support returning the product names, this PR retrieves the product names from the WooCommerce product data. When looping through the report results, we populate the name the first time the product is encountered. We use the helper functions both for retrieving the name and converting an MC ID to a WooCommerce product ID. The layout of this ID would be `gla_123` which converts to 123.

Note: The regular expression for parsing the MC ID was altered. Since it matches the end of the string it isn't necessary to include the separator `:` in the pattern. This allows us to reuse the same function for any of the following ID's: `gla_123` or `online:en:IE:gla_123`

Closes #455 
Unblocks #585

### Detailed test instructions:

1. Get some mocked report data possibly with a proxy service: https://github.com/woocommerce/google-listings-and-ads/pull/659
2. Make sure the mocked report data returns product ID's which are existing on the test site (leave some with non existing ids)
3. Send a request: `GET https://domain.test/wp-json/wc/gla/mc/reports/products?before=2021-03-14&after=2021-03-01&interval=day&fields[0]=clicks&fields[1]=impressions`
4. Confirm that the products contain both the ID and name (the ones that don't have an existing product will use the ID to populate the name field)

Example response:
```json
{
	"free_listings": null,
	"products": [
		{
			"id": "gla_44",
			"subtotals": {
				"clicks": 107,
				"impressions": 634
			},
			"name": "Durable Marble Bag"
		},
		{
			"id": "gla_48",
			"subtotals": {
				"clicks": 125,
				"impressions": 719
			},
			"name": "Synergistic Wooden Bag"
		},
		{
			"id": "gla_49",
			"subtotals": {
				"clicks": 107,
				"impressions": 112
			},
			"name": "Mediocre Wooden Shoes"
		},
		{
			"id": "gla_6789",
			"subtotals": {
				"clicks": 18,
				"impressions": 251
			},
			"name": "gla_6789"
		}
	],
	"intervals": [
		{
			"interval": "2021-05-01",
			"subtotals": {
				"clicks": 18,
				"impressions": 234
			}
		},
		{
			"interval": "2021-05-02",
			"subtotals": {
				"clicks": 89,
				"impressions": 400
			}
		},
		{
			"interval": "2021-05-03",
			"subtotals": {
				"clicks": 2,
				"impressions": 130
			}
		},
		{
			"interval": "2021-05-04",
			"subtotals": {
				"clicks": 123,
				"impressions": 589
			}
		},
		{
			"interval": "2021-05-05",
			"subtotals": {
				"clicks": 98,
				"impressions": 100
			}
		},
		{
			"interval": "2021-05-06",
			"subtotals": {
				"clicks": 9,
				"impressions": 12
			}
		},
		{
			"interval": "2021-05-07",
			"subtotals": {
				"clicks": 18,
				"impressions": 251
			}
		}
	],
	"totals": {
		"clicks": 357,
		"impressions": 1716,
		"sales": 0
	},
	"next_page": null
}
```


### Changelog Note:
* Fix - Add product names for Free Listing reports from WooCommerce product data.
